src/frontend/ocamlmerlin/new/new_merlin.ml

```diff
@@ -147,10 +147,15 @@ let run = function
         prerr_endline ("Exception: " ^ Printexc.to_string exn);
         1
 
-let restore_cwd old f args =
-  match f args with
-  | exception e -> Sys.chdir old; raise e
-  | res -> Sys.chdir old; res
+let with_wd ~wd ~old_wd f args =
+  try
+    Sys.chdir wd;
+    log ~title:"run" "changed directory to %S (old wd: %S)" wd old_wd;
+    Fun.protect ~finally:(fun () -> Sys.chdir old_wd) (fun () -> f args)
+  with _ ->
+    log ~title:"run" "cannot change working directory to %S (old wd: %S)"
+      wd old_wd;
```

Comment:
To avoid catching exceptions of `f` by mistake, I suggest to use `match with exception` instead:
```suggestion
let with_wd ~wd ~old_wd f args =
  match Sys.chdir wd with
  | () ->
    log ~title:"run" "changed directory to %S (old wd: %S)" wd old_wd;
    Fun.protect ~finally:(fun () -> Sys.chdir old_wd) (fun () -> f args)
  | exception Sys_error _ ->
    log ~title:"run" "cannot change working directory to %S (old wd: %S)"
      wd old_wd;
```

Reply:
Right, I will put your match back and then rebase and force puh to yoour branch if you don't mind